Add openSUSE/SLES support

tboerger opened the pull request:
In order to integrate openSUSE/SLES including the available patches for CVEs I have started to work on this pull request. I already opened this pull   request to get feedback early as possible. Below you can see a checklist for the remaining items.
- [x] Fetcher for openSUSE
- [x] Write tests for openSUSE fetcher
- [x] Fetcher for SLES
- [x] Write tests for SLES fetcher

Do I miss something? Should we abstract the shared functionality from RedHat and openSUSE/SLES?

Quentin-M commented:
@Hi,

This WIP is truly exciting. Thanks for contributing!
- Absolutely, we should extract the OVAL parser, I believe that [utils/](https://github.com/coreos/clair/tree/836d37b/utils) should be a good place for it. We would be able to test it independently
- We may need new [detectors](https://github.com/coreos/clair/tree/836d37b/worker/detectors) to fully support openSUSE/SLES in Clair

Out of curiosity, do you know how often openSUSE/SLES base images are used?

tboerger commented:
The official `opensuse:42.1` image is a pretty good example to demonstrate that it already works with the existing worker implementation.

tboerger commented:
> Out of curiosity, do you know how often openSUSE/SLES base images are used?

I really can't tell you how often it gets used. But at least for SLES it will increase for sure. I would like to hook Clair into [Portus](https://github.com/SUSE/Portus) for automated vulnerability checks.

tboerger commented:
We've got Hackweek now at SUSE and @jordimassaguerpla wants to finish this PR.

jordimassaguerpla commented:
I will refactor now by extracting common code first from sle and opensuse and then from redhat and suse.

jordimassaguerpla commented:
I am going to rebase all in one commit

jordimassaguerpla commented:
@Quentin-M : What do you think how it looks now? I am not very good with Go, so feel free to give me any advice . I still have tomorrow and can also work on that on my spare time next week.

thanks in advance

tboerger commented on updater/fetchers/oval/oval.go:
```diff
@@ -0,0 +1,306 @@
+// Copyright 2015 clair authors
```
That should be part of utils and not fetcher

tboerger commented on updater/fetchers/oval/oval.go:
```diff
@@ -0,0 +1,306 @@
+}
+
+var (
+	log = capnslog.NewPackageLogger("github.com/coreos/clair", "updater/fetchers/sle")
```
Sle? :)

tboerger commented on updater/fetchers/oval/oval.go:
```diff
@@ -0,0 +1,306 @@
+	return
+}
+
+func priority(def definition) types.Priority {
```
This function must be generic or distribution specific

tboerger commented:
The variable names are not really go-like

jordimassaguerpla commented:
can you be more concrete on the variable names?

tboerger commented:
> can you be more concrete on the variable names?

Sure, `OpenSUSE_Info` should be `OpenSUSEInfo`, `SLE_Info` should be `SLEInfo` and so on. Separation with underscores is AFAIK pretty uncommon in Go.

tboerger commented:
@jordimassaguerpla hopefully you can fix the remaining NITPICKing? Afterwards @Quentin-M should be happy for sure to merge it :)

tboerger commented:
@guypod what are you talking about?

Quentin-M commented:
@tboerger: @guypod is talking about #175, wrong thread.

guypod commented:
Oops, my bad  - too many similar looking GitHub tabs open ;)
Sorry for the noise!

Deleted my comment to avoid future confusion.

jordimassaguerpla commented:
sure. Sorry for the delay, hackweek ended and I didn't find the time to do it.I was refactoring the rhel code to use the oval parser and that is why I had not sent it yet.  I'll do it asap.

jordimassaguerpla commented:
Hi! I just refactored everything. I extracted the oval parser from rhel and use that for suse and opensuse.
Since commiting here would be a mess, I created another PR

https://github.com/coreos/clair/pull/218

Closing this one.

jordimassaguerpla commented:
mm. I think can't close this one. Thomas can you?

tboerger commented:
Why closing this one? Just rebase and squash commits? :)

jordimassaguerpla commented:
ok :-)

Branch rebased. I am closing the other one

jordimassaguerpla commented on updater/fetchers/rhel/rhel.go:
```diff
@@ -34,326 +29,98 @@ const (
-	if err != nil {
-		return resp, err
-	}
-	firstRHSA, err := strconv.Atoi(flagValue)
```
The tests pass but this thing of getting the firstRHSA from the flagValue ... not sure If I extracted this correctly or if I even understood what this was for .... Can someone else just take a look and see if I missed something here? Thanks and sorry I if did.

Quentin-M replied:
This ensures that Clair doesn't try to re-parse RHSAs that have already been processed and stored - in order to make the updates faster.

jordimassaguerpla commented:
I am copying here also the comments from the other pull request .... sorry for the noise

jordimassaguerpla commented:
I am testing opensuse with this docker image
https://gist.github.com/jordimassaguerpla/8ce11ed97284b4f4a5b40647b1730b7b

I tried it with opensuse by starting clair and then running

./analyze-local-images -minimum-severity Unknown opensuse_vul

opensuse_vul is the opensuse docker image that contains at least one vulnerability (see previous comment).

The result is as expected, it outputs the expected vulnerability (and others), as long as you specify the severity "Unknown". That is something we are working on too.

tboerger commented:
I'm working on getting severity informations for SUSE provided. It will be hopefully an additional field like at https://github.com/openSUSE/clair/blob/master/updater/fetchers/rhel/testdata/fetcher_rhel_test.1.xml#L29 so you should check this field first and only if that doesn't match fall back to the title substring.

jordimassaguerpla commented:
https://github.com/coreos/clair/pull/199/files#diff-1ff0ffc3d0f5219b7b96158670ea9a3eR324 

^ I am already checking for this field. Only when it is not found it is defaulted to Unknown.

tboerger commented:
But you are parsing the severity from the title, I say this should only be a fallback if the specific tag doesn't exist

jordimassaguerpla commented:
@tboerger : true.

But that is how it was being done so far for rhel. I just refactored and reused that same code as much as I could, to keep it simple/readable.

If you take a look, you will see that there is a similar issue with parsing the OS version or the package name and version: they all come from a comment field while there are specific elements for that, which would be better and less error prone to use.

What I mean is, can we create a new PR which improves all of that and keep this just as a refactoring? I think this is already introduces a lot of changes in existing code (rhel -> oval).

Then we can create a PR for better parsing the oval elements.

What do you think?

jordimassaguerpla commented:
@Quentin-M : Hi! Any news on this :) ? Thanks in advance

jzelinskie commented on utils/oval/oval.go:
I don't think this error check is necessary

tboerger replied:
True :D

jzelinskie commented on utils/oval/oval.go:
remove this newline -- error checks should always be immediately following a call that returns an error

jzelinskie commented on utils/oval/oval.go:
newlines below each function signature

jzelinskie commented on utils/oval/oval.go:
missing a space before `Namespace`

also comments need to be full sentences (including punctuation) and wrapped to 80char lines

jzelinskie commented on utils/oval/oval.go:
remove this newline

jzelinskie commented on utils/oval/oval.go:
rm this \n

jzelinskie commented on utils/oval/oval.go:
You can remove this whole for loop and replace it with append with a splat

``` go
resp.Vulnerabilities = append(resp.Vulnerabilities, vs...)
```

jzelinskie commented on utils/oval/oval.go:
do we want a space or something between the distname and the word updater?

jordimassaguerpla replied:
tbh, I am not sure what is the FlagName for, I just extracted the code from the rhel parser ... Do you know it?

jordimassaguerpla replied:
I see it is being used in the updater/updater.go ...
in general I would try to not use spaces in a flag name. What about a "_" ?

jzelinskie replied:
Sounds good to me

jzelinskie commented on utils/oval/oval.go:
This is a very ambiguous log message

jzelinskie commented on utils/oval/oval.go:
before this line we could check to see if we have both a valid featureVersion and osVersion and break out of the loop

jzelinskie commented on utils/oval/oval.go:
this is a very ambiguous log message

jzelinskie commented on utils/oval/oval.go:
Is the value of %v for criterions useful? Is there a useful identifier for each section of criterions so that someone could refer to the file and know where the failure was?

jordimassaguerpla replied:
Yes, there is the test_ref that can be used for looking into the xml file to know which criterion was.

jzelinskie commented on utils/oval/oval.go:
Similar to the other recommendation you can use 1 append and the splat

``` go
featureVersionParametersArray = append(featureVersionParametersArray, featureVersionParamters...)
```

Is there a reason you're explicitly making a copy here?

jzelinskie commented on utils/oval/oval.go:
rm this \n

jzelinskie commented on utils/oval/oval.go:
the old splat trick could be used here too

jzelinskie commented on utils/oval/oval.go:
rm this \n

jzelinskie commented on utils/oval/oval.go:
could use the splat append here too

jzelinskie commented on utils/oval/oval.go:
another splat append here

jzelinskie commented:
Thanks for the PR! This is a really useful refactor of the oval stuff into a shared package. I mostly commented on code that was already there, but simply moved.

jordimassaguerpla commented:
@jzelinskie : I've done all the changes you suggested.
Is there anything else?
I am leaving the commits for your revision but I will rebase once I get the "lgtm" from you or @Quentin-M

jzelinskie commented:
Just that last tiny thing of changing the updater flag to have an underscore and then it LGTM! Thanks a bunch for all your effort, this change is awesome!

tboerger commented:
Cool to see it's nearby getting merged :clap:

jordimassaguerpla committed:
Integrated a fetcher for openSUSE and for SUSE Linux Enterprise

We extracted oval parser from rhel and used that for opensuse and
SUSE Linux Enterpise

Signed-off-by: Thomas Boerger <tboerger@suse.de>
Signed-off-by: Jordi Massaguer Pla <jmassaguerpla@suse.de>

jordimassaguerpla commented:
I just did the change on the updated flag and rebased everything so it looks cleaner when you merge. Thanks a lot for taking the time to review it :) ! I am very excited to be contributing :) !

arjunm183 commented:
Is this completed to integrate clair with SLES Updater.

jzelinskie commented:
This change actually got reverted because I couldn't ever get an answer from SUSE's legal department about the license of the Vulnerability data.

flavio commented:
Hello from SUSE, I'm a member of the team that wrote this PR. What kind of legal clearance are you looking for and who have you contacted?

Pinging also @msmeissn from SUSE's security team. He might be able to help with your questions about the usage of the vulnerability data.

jzelinskie commented:
Hi @flavio 
I simply need to know how the vulnerability data is licensed to be used. I sent an email to security@ and received a response from Marcus, but never got a follow up.

msmeissn commented:
I am sorry if I failed to answer, but I got the answer from my legal department.

https://www.suse.com/de-de/support/security/oval/

The OVAL data is provided by SUSE under the Creative Commons License 4.0 with Attribution for Non-Commercial usage (CC-BY-NC-4.0).

I hope this is satisfying.

Rots commented:
@jzelinskie, is this expected to be merged again at some point now?

flavio commented:
@Rots clair's code changed and merging the reverted code again is not possible. I'm working on that and I'll file a PR in the next weeks (I'm a bit busy ATM).

LJosef commented:
@flavio looking forwarding to the progress on this merging. Great appreciate!

chennanfei commented:
Looking forwarding to the progress, too!

Payback159 commented:
Looking forwarding to the progress, too!

maorgo commented:
Hi,
so is open suse supported for Clair yet?

Thanks

flavio commented:
@maorgo not yet, you have to follow https://github.com/coreos/clair/pull/503 and https://github.com/coreos/clair/pull/506
